### PR TITLE
Prefs[TASK[28635] Create UI (Check In Screen)

### PR DESCRIPTION
**Pull request by hieuna-441907**
## Related Tickets
- [#Task 28635: Create UI (Check In Screen)](https://edu-redmine.sun-asterisk.vn/issues/28635)

## WHAT
- Create UI (Check In Screen)
## HOW
- Create UI (Check In Screen)
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (HuyTQ)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/ui/adapter/CheckInViewPagerAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    private var fragments = mutableListOf<Fragment>()
+    private var fragmentTitles = mutableListOf<String>()
+
+    override fun getItem(position: Int) = fragments[position]
+
+    override fun getCount() = fragments.size
+
+    fun updateList(fragments: MutableList<Fragment>) {
+        this.fragments.clear()
+        this.fragments = fragments
```
Comment (quockhanhng) [inline app/src/main/java/com/sunasterisk/travelapp/ui/adapter/CheckInViewPagerAdapter.kt]:
```suggestion
        this.fragments.run {
            clear()
            addAll(fragments)
        }
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/checkin/BookRoomFragment.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+    override val layoutResource get() = R.layout.fragment_book_room
+
+    override fun initComponents() {
+
+    }
+
+    companion object {
+        const val TITLE = "Select Room"
+        fun newInstance() = BookRoomFragment()
```
Comment (quockhanhng) [inline app/src/main/java/com/sunasterisk/travelapp/ui/checkin/BookRoomFragment.kt]: sao không gọi thẳng từ main constructor anh

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/BackdropViewPager.kt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    override fun onClick(view: View) {
+        when (view.id) {
+            R.id.imageViewStep -> slideFrontLayout()
+            R.id.materialButtonContinue -> onNextPage()
+        }
+        if (view == toolbarNavigation) {
+            handBackPress()
+        }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/BackdropViewPager.kt:71]:
```suggestion
        when (view.id) {
            R.id.imageViewStep -> slideFrontLayout()
            R.id.materialButtonContinue -> onNextPage()
            toolbarNavigation?.id -> handBackPress()
        }
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/BackdropViewPager.kt [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    private fun onNextPage() {
+        if ((currentPos < checkInViewPagerAdapter.count - 1)) {
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/BackdropViewPager.kt]:
```suggestion
        if (currentPos >= checkInViewPagerAdapter.count - 1) return
```

**Review by aug16vcc — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/sunasterisk/travelapp/ui/checkin/BookRoomFragment.kt (hunk 2 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/ui/checkin/BookRoomFragment.kt]: ok em, chỗ này bị thừa

File: app/src/main/java/com/sunasterisk/travelapp/ui/adapter/CheckInViewPagerAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import androidx.fragment.app.FragmentStatePagerAdapter
+
+class CheckInViewPagerAdapter(
+    fm: FragmentManager
+) : FragmentStatePagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
+
+    private val fragments = mutableListOf<Fragment>()
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/adapter/CheckInViewPagerAdapter.kt:11]: T không chắc là có đc giữ `Fragment` không? Có thể bị `memory-leak`. Cái này hỏi Huy nhé, t cũng muốn biết

File: app/src/main/java/com/sunasterisk/travelapp/ui/checkin/CheckInActivity.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        toolbarCheckIn.post {
+            viewPagerMainCheckInContent.setToolbar(toolbarCheckIn)
+        }
+        materialButtonContinue.post {
+            materialButtonContinue.setOnClickListener(viewPagerMainCheckInContent)
+        }
+    }
+
+    companion object {
+        val listFragments = mutableListOf<Fragment>(
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/checkin/CheckInActivity.kt]:
```suggestion
        val listFragments = listOf(
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/checkin/CheckInActivity.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            TravelerInforFragment(),
+            BookRoomFragment(),
+            ConfirmFragment()
+        )
+
+        val listFragmentTitle = mutableListOf(
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/checkin/CheckInActivity.kt]:
```suggestion
        val listFragmentTitle = listOf(
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/custom/SwipeControllableViewPager.kt [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+class SwipeControllableViewPager constructor(
+    context: Context,
+    attrs: AttributeSet
+) : ViewPager(context, attrs) {
+    var swipeEnable = false
+
+    init {
+        val scroller = ViewPager::class.java.getDeclaredField("mScroller")
+        scroller.isAccessible = true
+        scroller.set(this, FastOutSlowInScroller(context, FastOutSlowInInterpolator()))
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/custom/SwipeControllableViewPager.kt]:
```suggestion
        val scroller = ViewPager::class.java.getDeclaredField("mScroller")
        try {
            scroller.isAccessible = true
            scroller.set(this, FastOutSlowInScroller(context, FastOutSlowInInterpolator()))
        } finally {
            scroller.isAccessible = false
        }
```

Comment (aug16vcc) [reply app/src/main/java/com/sunasterisk/travelapp/ui/custom/SwipeControllableViewPager.kt]: Thêm nữa là cái `mScroller` là hard code nhé

File: app/src/main/java/com/sunasterisk/travelapp/utils/ViewExt.kt [hunk 9]
```diff
@@ -29,3 +31,14 @@ fun ImageView.setImageUrl(url: String) {
...
+fun Toolbar.getToolbarNavigationButton(): ImageButton? {
+    val size = this.childCount
+    (0 until size).forEach {
+        val child = this.getChildAt(it)
+        if (child is ImageButton) {
+            return child
+        }
+    }
+    return null
+}
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/utils/ViewExt.kt]:
```suggestion
fun Toolbar.getToolbarNavigationButton() =
    (0 until this.childCount).asSequence()
         .map(this::getChildAt)
         .firstOrNull { it is ImageButton }
```
Thử thế này xem

File: app/src/main/java/com/sunasterisk/travelapp/ui/adapter/CheckInViewPagerAdapter.kt (hunk 5 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/ui/adapter/CheckInViewPagerAdapter.kt:11]: ok ông, t truyền vào supportFragmentManager, check thì thấy configuration change, onDetach của bọn child fragments ko chạy, trong phạm vi bọn t set màn hình hướng dọc thôi

File: app/src/main/java/com/sunasterisk/travelapp/utils/ViewExt.kt (hunk 9 above)
Comment (hieuna-441907) [reply app/src/main/java/com/sunasterisk/travelapp/utils/ViewExt.kt]: ok ông, gọn hơn rất nhiều

**Pull request merged**